# HG changeset patch
# User Rik <rik@octave.org>
# Date 1531332683 25200
#      Wed Jul 11 11:11:23 2018 -0700
# Branch stable
# Node ID 18148dc92122adaf6cbce1301712a137f891fb93
# Parent  9f422e49be152a88e8a7ac9f8258a30a5b30b037
Fix regression in 4.2.0 which disallowed printing with some Ghostscript devices.

* __print_parse_opts__.m: New boolean variable "unknown_device" initially
set to true.  Set variable to false if search in dev_list (devices),
gs_aliases, or gs_device_list is succesful.

diff --git a/scripts/plot/util/private/__print_parse_opts__.m b/scripts/plot/util/private/__print_parse_opts__.m
--- a/scripts/plot/util/private/__print_parse_opts__.m
+++ b/scripts/plot/util/private/__print_parse_opts__.m
@@ -249,23 +249,25 @@ function arg_st = __print_parse_opts__ (
     ## Allow tests when no figures are present.
     __graphics_toolkit__ = get (0, "defaultfigure__graphics_toolkit__");
   endif
 
   if (strcmp (__graphics_toolkit__, "gnuplot"))
     suffixes(strncmp (dev_list, "epslatex", 8)) = {"eps"};
   endif
 
+  unknown_device = true;
+  default_suffix = "";
   match = strcmpi (dev_list, arg_st.devopt);
-  if (! any (match))
-    error ("print: unknown device %s", arg_st.devopt);
+  if (any (match))
+    unknown_device = false;
+    default_suffix = suffixes{match};
   endif
-  default_suffix = suffixes{match};
 
-  if (dot == 0 && ! isempty (arg_st.name))
+  if (dot == 0 && ! isempty (arg_st.name) && ! isempty (default_suffix))
     arg_st.name = [arg_st.name "." default_suffix];
   endif
 
   if (arg_st.append_to_file)
     if (isempty (arg_st.name))
       arg_st.append_to_file = false;
     elseif (any (strcmpi (arg_st.devopt, {"eps", "eps2", "epsc", "epsc2", ...
                                           "ps", "ps2", "psc", "psc2", "pdf"})))
@@ -292,40 +294,46 @@ function arg_st = __print_parse_opts__ (
 
   if (any (strcmp (arg_st.devopt, {"ps", "ps2", "psc", "psc2", "pdf"})))
     arg_st.formatted_for_printing = true;
   endif
 
   aliases = gs_aliases ();
   if (any (strcmp (arg_st.devopt, fieldnames (aliases))))
     arg_st.devopt = aliases.(arg_st.devopt);
+    unknown_device = false;
   endif
 
   if ((any (strcmp (arg_st.devopt, gs_device_list))
        && ! arg_st.formatted_for_printing)
       || any (strcmp (arg_st.devopt, {"pswrite", "ps2write", "pdfwrite"})))
+    unknown_device = false;
     ## Use ghostscript for graphic formats
     arg_st.ghostscript.device = arg_st.devopt;
     arg_st.ghostscript.output = arg_st.name;
     arg_st.ghostscript.antialiasing = true;
     if (arg_st.formatted_for_printing)
       arg_st.ghostscript.epscrop = ! arg_st.loose;
     else
       ## pstoedit throws errors if the EPS file isn't cropped
       arg_st.ghostscript.epscrop = true;
     endif
   elseif (all (! strcmp (arg_st.devopt, dev_list)))
-    ## Assume we are formating output for a printer
+    ## Assume we are formatting output for a printer
     arg_st.formatted_for_printing = true;
     arg_st.ghostscript.device = arg_st.devopt;
     arg_st.ghostscript.output = arg_st.name;
     arg_st.ghostscript.antialiasing = false;
     arg_st.ghostscript.epscrop = ! arg_st.loose;
   endif
 
+  if (unknown_device)
+    error ("print: unknown device %s", arg_st.devopt);
+  endif
+
   if (arg_st.send_to_printer)
     if (isempty (arg_st.name))
       ## Pipe the ghostscript output
       arg_st.name = "-";
     else
       error ("print: a filename may not specified when spooling to a printer");
     endif
     if (! any (strcmp (arg_st.devopt, gs_device_list)))
